### PR TITLE
GLT-2201: Load Sequencer Workflow

### PR DESCRIPTION
**Pull request by djcooke**
Will just squash this into one commit when merging

Note: There are no links to the workflow yet, but you can access it via `.../miso/workflow/new/LOAD_SEQUENCER` or edit an in-progress workflow via `.../miso/workflow/edit/{progressId}`

### CODE REVIEW

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java [hunk 1]
```diff
@@ -24,47 +44,282 @@ protected WorkflowName getWorkflowName() {
...
-  public WorkflowStepPrompt getStep(int stepNumber) {
-    return null;
+  private int asPartitionIndex(int stepNumber) {
+    return stepNumber - (containerStep.isExistingContainer() ? 1 : 2);
   }
 
   @Override
   public boolean isComplete() {
-    return false;
+    return partitionSteps.size() > 0 && Stream.concat(Stream.of(containerStep), partitionSteps.stream()).allMatch(this::hasInput);
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java]: MINOR Use isEmpty() to check whether the collection is empty or not. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java [hunk 2]
```diff
@@ -24,47 +44,282 @@ protected WorkflowName getWorkflowName() {
...
+    for (int i = 0; i < model.getPartitionCount(); ++i) {
+      partitionSteps.add(new PartitionStep(i, model.getPlatformType()));
+    }
   }
 
   @Override
   public void cancelInput() {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (partitionSteps.stream().anyMatch(this::hasInput)) {
+      getLastCompletedPartitionStep().cancelInput();
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java]: MAJOR A "NullPointerException" could be thrown; "getLastCompletedPartitionStep()" can return null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java [hunk 3]
```diff
@@ -24,47 +44,282 @@ protected WorkflowName getWorkflowName() {
...
+    }
+  }
+
+  private static class ContainerStep implements WorkflowStep {
+
+    private static final WorkflowStepPrompt PROMPT = new WorkflowStepPrompt(Sets.newHashSet(InputType.SEQUENCER_PARTITION_CONTAINER,
+        InputType.STRING),
+        "Scan the Sequencing Container serial number (SN)");
+
+    private SequencerPartitionContainerProgressStep containerStep;
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java]: MAJOR Rename field "containerStep" [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1700)

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 4]
```diff
@@ -84,29 +85,54 @@ var WorkflowDisplay = (function() {
...
+      });
+      table.append(makeRow(rowNum++, [arrowCell, messageCell, actionCell]));
+    }
 
-    for (var i = 0; i < logEntries.length; i++) {
-      table.prepend(jQuery("<tr>").append(
-          [makeArrowCell(i, currentStepNumber), jQuery("<td>").text(logEntries[i]), jQuery("<td>"), makeModifyCell(i, logEntries.length)])
-          .click(makeLogEntryClickHandler(i, logEntries.length, workflowId)));
+    for (var i = logEntries.length - 1; i >= 0; i--) {
+      var arrowCell = makeArrowCell(i, currentStepNumber);
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: MAJOR Rename "arrowCell" as this name is already used in declaration at line 123. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 5]
```diff
@@ -84,29 +85,54 @@ var WorkflowDisplay = (function() {
...
+      var messageCell = jQuery("<td>").text(logEntries[i]);
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: MAJOR Rename "messageCell" as this name is already used in declaration at line 124. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 6]
```diff
@@ -84,29 +85,54 @@ var WorkflowDisplay = (function() {
...
+      var actionCell = makeActionCell(i, logEntries.length, workflowId)
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: MAJOR Rename "actionCell" as this name is already used in declaration at line 125. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)

**Comment by sqwci** [comment]
SonarQube analysis reported 20 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultContainerService.java#L121](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultContainerService.java#L121): Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [MenuController.java#L219](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/MenuController.java#L219): Define a constant instead of duplicating this literal "/login.jsp" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [MenuController.java#L236](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/MenuController.java#L236): Define a constant instead of duplicating this literal "error" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![MAJOR][MAJOR] [DefaultRunService.java#L571](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L571): Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [DefaultRunService.java#L685](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L685): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)
1. ![MINOR][MINOR] [DefaultContainerService.java#L85](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultContainerService.java#L85): Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.service.security.AuthorizationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultRunService.java#L172](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L172): Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.service.security.AuthorizationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultRunService.java#L179](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L179): Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.service.security.AuthorizationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultRunService.java#L186](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L186): Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.service.security.AuthorizationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultRunService.java#L587](https://github.com/TGAC/miso-lims/blob/ed61b52b67ecace5a4559e15582264ff14fa539b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultRunService.java#L587): Replace this lambda with a method reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1612)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/webapp/styles/images/arrow-back.svg [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	<g id="_x31_">
+		<g>
+			<path d="M283.123,159.091V25.423c0.771-6.686-1.065-13.598-6.232-18.743c-8.975-8.907-23.524-8.907-32.5,0L18.03,254.889
+				c-4.782,4.759-6.822,11.06-6.504,17.292c-0.317,6.232,1.722,12.533,6.504,17.292l225.137,246.849
+				c8.408,7.184,23.819,11.264,33.746,1.359c5.145-5.145,7.388-9.473,6.618-16.158c0,0,0-122.586,0-135.98
+				c149.578,0,284.855,107.695,311.711,249.432c18.085-41.564,28.238-87.344,28.238-135.572
+				C623.48,311.457,471.093,159.091,283.123,159.091z"/>
+		</g>
+	</g>
+</g>
```
Comment (apmasell) [inline miso-web/src/main/webapp/styles/images/arrow-back.svg]: I think it's weird that all of these SVGs have a bunch of empty groups at the end. I'm not sure why Illustrator does this. Consider deleting them.

**Pull request merged**